MAD.Extensions.EFCore/UpsertExtensions.cs

```diff
@@ -11,32 +11,51 @@ namespace MAD.Extensions.EFCore
 {
     public static class UpsertExtensions
     {
-        public static void Upsert(this DbContext dbContext, object entity, Action<object> transformations = null)
+        public static void Upsert(this DbContext dbContext, object entity)
         {
             dbContext.ChangeTracker.TrackGraph(entity, g =>
             {
-                var entity = g.Entry.Entity;
                 var entityType = g.Entry.OriginalValues.EntityType;
 
                 if (entityType is null)
                     return;
 
-                transformations?.Invoke(entity);
+                if (g.InboundNavigation is null == false && g.InboundNavigation.IsCollection())
+                {
+                    var principalKeyProperties = g.InboundNavigation.ForeignKey.PrincipalKey.Properties;
+                    var dependantKeyProperties = g.InboundNavigation.ForeignKey.Properties;
+
+                    var pkValues = new List<object>();
+
+                    foreach (var pk in principalKeyProperties)
+                    {
+                        var value = dbContext.Entry(entity).Property(pk.Name).CurrentValue;
```

Comment:
Won't this cause issue with deeply nested entity graphs? Entity is the root object. Shouldn't this be g.Entry.Entity as it was before?

Reply:
I've set this to use the entry entity now